internal/fanoutconsumer/metrics.go

```diff
@@ -18,36 +18,22 @@ import (
 // NewMetrics wraps multiple metrics consumers in a single one.
 // It fanouts the incoming data to all the consumers, and does smart routing:
 //   - Clones only to the consumer that needs to mutate the data.
-//   - If all consumers needs to mutate the data one will get the original data.
+//   - If all consumers needs to mutate the data one will get the original mutable data.
 func NewMetrics(mcs []consumer.Metrics) consumer.Metrics {
-	if len(mcs) == 1 {
-		// Don't wrap if no need to do it.
-		return mcs[0]
-	}
-	var pass []consumer.Metrics
-	var clone []consumer.Metrics
-	for i := 0; i < len(mcs)-1; i++ {
-		if !mcs[i].Capabilities().MutatesData {
-			pass = append(pass, mcs[i])
+	mc := &metricsConsumer{}
+	for i := 0; i < len(mcs); i++ {
+		if mcs[i].Capabilities().MutatesData {
+			mc.mutable = append(mc.mutable, mcs[i])
 		} else {
-			clone = append(clone, mcs[i])
+			mc.readonly = append(mc.readonly, mcs[i])
 		}
 	}
-	// Give the original data to the last consumer if no other read-only consumer,
-	// otherwise put it in the right bucket. Never share the same data between
-	// a mutating and a non-mutating consumer since the non-mutating consumer may process
-	// data async and the mutating consumer may change the data before that.
-	if len(pass) == 0 || !mcs[len(mcs)-1].Capabilities().MutatesData {
-		pass = append(pass, mcs[len(mcs)-1])
-	} else {
-		clone = append(clone, mcs[len(mcs)-1])
-	}
-	return &metricsConsumer{pass: pass, clone: clone}
+	return mc
 }
 
 type metricsConsumer struct {
-	pass  []consumer.Metrics
-	clone []consumer.Metrics
+	mutable  []consumer.Metrics
+	readonly []consumer.Metrics
 }
 
 func (msc *metricsConsumer) Capabilities() consumer.Capabilities {
```

Comment:
This is a bug. Can we fix this first? Since data are mutated with current implementation for example if we have 2 mutable consumers.

Reply:
Should we consider changing the state a mutation? The original data isn't touch. Anyway, it I don't think it should matter because the Capabilities of the fanout consumer are not being used in the components graph building process if I'm not mistaken.

Reply:
We need to check that :)

Reply:
Sure. We add a capability processor in front of processors [here](https://github.com/open-telemetry/opentelemetry-collector/blob/53615832e65416a89667611481b8765ab244021c/service/internal/graph/graph.go#L291-L300) without looking at the fanout consumer at the end of the processors' chain. When building connector, [we look at capability nodes](https://github.com/open-telemetry/opentelemetry-collector/blob/53615832e65416a89667611481b8765ab244021c/service/internal/graph/nodes.go#L241-L258) exposed on the receiver side and wrap the fanout consumer with capability consumer based on that. The regular receiver use case is straight forward. So I don't see any place where capabilities of the fanout consumer is used.